az-digital/az_quickstart
# Fixes #1537 Distribution Config Export Commands

## Description
One task frequently during development is the re-export of module owned configuration to snapshot changes made in the development environment. In particular this would be desirable after the transition to D10 to catch configuration drift. This PR introduces functionality that examines module-owned configuration items and attempts to re-export changes in the active configuration to the module files.

A command is also added to output a single configuration item with the site-specific hashes removed, since Drupal console is no longer available in our D10 build. (The drush equivalent does not remove the site-specific hashes)

- `drush az-core-distribution-config`
- `drush az-core-config-export-single`

These are development tools to assist with exporting distribution configuration.

**NOTE: our configuration contains a few edge cases that are not currently handled by this PR but I think would make good future improvements in functionality:**

- Our permission list in roles contain all applicable permissions. Missing active permissions will be removed when re-export. Ideally, this functionality should merge the inactive permissions so they are not lost - i.e. It would be better if only available permissions that were explicitly removed were removed in the export. Take care when re-exporting roles.
- New dependent config is not export, only files already present. It would be nice if dependent configuration was export if it's not already in another module file.
- Quickstart override config will be re-export to the non-overridden version, e.g. metatag-enabled settings. It would be good if this functionality only exported to the overridden copy e.g. `az_seo`

I hopefully intend to handle the above cases in future improvements to this initial functionality.

## Related issues
#1537 

## How to test

- `lando drush az-core-config-export-single configuration.item.name`, e.g. `lando drush az-core-config-export-single cas.settings`
- verify the chosen configuration is output to the console
- `lando drush az-core-distribution-config` or `lando drush az-core-distribution-config az_module1,az_module2,az_module3` with a comma delimited list of module names
- step through the process of confirming which items should be export
- verify that any changes were written to the module config folders

See description for cases which are not currently handled by this initial command (further work or future PRs).

## Types of changes


### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
- **Minor release changes**
- **Other or unknown**

### Drupal core
- **Patch release changes**
- **Minor release changes**
- **Other or unknown**

### Drupal contrib projects
- **Patch release changes**
- **Minor release changes**
- **Other or unknown**

## Checklist
